#!/usr/bin/perl
# Base algorithm adapted from: http://genomics-array.blogspot.com/2011/02/calculating-n50-of-contig-assembly-file.html
# Edit 9/21/2016 -- Added option to calculate NG50 instead of N50 and option to read fai files
# Edit 1/26/2017 -- Added option to calculate N[G]50 from a bed file of gap regions (like those generated by GetMaskBedFasta.jar)


use strict;
use Getopt::Std;

my %opts;

my $usage = "perl $0 -f <input fasta file> OR -i <input fai file> OR -b <input bed file from GetMaskBedFasta.jar>
	-f	Input fasta file
	OR
	-i	Input samtools fai file
	OR
	-b	Input gap bed file
	
	[optional]
	-g	Genome size (for NG50 calculation)
	-s	Gap size cutoff (for gap bed input; assumed 3bp cutoff)\n";
getopt('figbs', \%opts);

my $gapThresh = (defined($opts{'s'}))? $opts{'s'} : 3; 
my $useBed = 0;
if(!defined($opts{'f'}) && !defined($opts{'i'}) && !defined($opts{'b'})){
	print $usage;
	exit;
}

my $IN;
my $fasta = 0;

if(defined($opts{'f'})){
	open($IN, "< $opts{f}") || die "Could not open input fasta file: $opts{f}!\n";
	$fasta = 1;
}elsif(defined($opts{'i'})){
	open($IN, "< $opts{i}") || die "Could not open input fai file: $opts{i}!\n";
}elsif(defined($opts{'b'})){
	open($IN, "< $opts{b}") || die "Could not open input gap bed file: $opts{b}!\n";
	$useBed = 1;
}

my $ng50 = 0;
my $totalLength = 0; 
my @arr;
if(defined($opts{'g'})){
	$ng50 = 1;
	$totalLength = $opts{'g'};
}

## Read Fasta File and compute length ###
if(defined($opts{'f'})){
	my $length = 0;	
	while(my $line = <$IN>){
		chomp $line; 
		if($line =~ /^>/){
			push (@arr, $length);
			unless($ng50){
				$totalLength += $length;
			}
			$length = 0;
			next;
		}
		$length += length($line);
	}
}elsif(defined($opts{'i'})){
	# Read FAI file and compute length
	while(my $line = <$IN>){
		chomp $line;
		my @segs = split(/\t/, $line);
		push(@arr, $segs[1]);
		unless($ng50){
			$totalLength += $segs[1];
		}
	}
}elsif(defined($opts{'b'})){
	# Read bed file and determine the largest length
	my $lastChr = "NA"; my $lastEnd = 0; my $lastStart = 0; my $added = 0;
	while(my $line = <$IN>){
		chomp $line;
		my @segs = split(/\t/, $line);
		if($segs[0] ne $lastChr){
			# New chromosome! add lengths if we didn't already first
			if(!$added){
				push(@arr, $lastEnd)
			}
			$lastEnd = 0;
			$lastChr = $segs[0];
		}
		
		if($segs[2] - $segs[1] >= $gapThresh){
			push(@arr, $segs[1] - $lastEnd);
			unless($ng50){
				$totalLength += $segs[1] - $lastEnd;
			}
			$lastEnd = $segs[2]; 
			$added = 1;
		}
	}
}
				

close($IN);

my @sort = sort {$b <=> $a} @arr;
my $n50; 
my $L50;
foreach my $val(@sort){
	$n50+=$val;
	$L50++;
	if($n50 >= $totalLength/2){
		my $type = ($ng50)? "G50" : "50";
		print "N$type length:\t$n50\n";
		print "N$type value:\t$val\n";
		print "L$type value:\t$L50\n";
		last; 
	}
}

exit;
